### DISCUSSION TITLE
How to add name to rows when doing an append join

### DISCUSSION THREAD

**Discussion by frosencrantz** [saulpw/visidata#1335] [Q&A]
I encountered a new question with some data I'm working with.  This is similar to my question in https://github.com/saulpw/visidata/discussions/1331 .

Now I have my data split across multiple CSV files, with one file per student, all the same column names.  I want to load all the data files into Visidata, and then do an append join.  But I want to make sure each lines reports the student name, in this case the file name is a good proxy for the student's name.

A.csv:
```csv
test,score
1,80
2,90
3,95
```
B.csv:
```csv
test,score
1,90
2,90
3,90
```

C.csv:
```csv
test,score
1,100
2,85
3,93
```

So I end up with the data labeled and combined, like below.  What is a good way to have Visidata label the data with the filename so I see the source of each row.

```
student,test,score
a,1,80
b,1,90
c,1,100
a,2,90
b,2,90
c,2,85
a,3,95
b,3,90
c,3,93
```

**Comment by ghost** [comment]
I couldn't think of a clean visidata version that didn't involve writing a custom command in Python, but I got nerd-sniped, and had to script a Zsh version:

``` zsh
#!/usr/bin/env zsh

output_file=combined
echo 'student,test,score' > "$output_file"
for file in *.csv; do
        # zsh magic to remove the file extension
	filename=($file(:r)) 
        # sed magic to add the filename and a comma to the start of all lines except the first line.
	sed -n -e "2,$ s/^/"$filename",/gp" "$file" >> "$output_file"
done
# if we use the csv extension from the start, it gets picked up by the for loop...
mv "$output_file" "$output_file".csv

```

**Comment by frosencrantz** [comment]
Thanks David!

I should have mentioned I do something similar as a workaround in the shell, but wanted a simple way to do it within Visidata.    I like your use of sed and zsh filename modifications.   I'm using a favorite tool of [aborruso](https://github.com/aborruso): https://github.com/johnkerl/miller 

I use the miller put and cat verbs.

**Comment by aborruso** [comment]
@frosencrantz you are a visidata master. But why do not use a visidata macro in which for each sheet, a column with the result of expression `=sheet` is added?
And then you will have this column in the sheets you want to append

**Comment by frosencrantz** [comment]
Thanks, @aborruso I was trying to use the native Visidata without diving into the Python.   You are right it would need something that can add a  column to all the selected sheets in an IndexSheet of `sheet.name` .  I could be the master at copy-pasta.

I don't know enough vd how to write a simple command or function that would add a column to selected sheets in an IndexSheet/SheetsSheet that would work within the context of each sheet, so it would be possible to add that `sheet.name` column.  That last part is what I don't know.


P.S. David I hope you are good.  I see your account is now listed as deleted. I hope we'll see you again. Thanks again for your suggestion.

**Comment by aborruso** [comment]
Hi @frosencrantz the way I'm able to do it is this:

Starting from a folder with my csv files and this vd

```
sheet	col	row	longname	input	keystrokes	comment
global		null_value	set-option			
	score		addcol-expr	sheet	=	create new column from Python expression, with column names as variables
```

I run

```bash
mkdir -p ./out
for i in *.csv;do name=$(basename "$i" .csv); vd -y -b -p fro.vd "$i" -o ./out/"$name".csv;done
```

And I have in output all the csv files with the new sheet name column.

And then I run `vd -y -b -p fro-m.vd ./out/*.csv -o out.tsv`, using this vd file

```
sheet	col	row	longname	input	keystrokes	comment
global		null_value	set-option			
			sheets-stack		Shift+S	open Sheets Stack: join or jump between the active sheets on the current stack
sheets			select-cols-regex	\.csv	g|	select rows matching regex in any visible column
sheets			join-selected	append	&	merge selected sheets with visible columns from all, keeping rows according to jointype
```

In output I have

```
test	score	sheet
1	80	a
2	90	a
3	95	a
1	90	b
2	90	b
3	90	b
1	100	c
2	85	c
3	93	c
```
